### PR TITLE
V2 client and deploy changes

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR adds functionality for creating a V2 client using the V1 interface and makes some changes to the V2 deployment process. Some V1 functionality is not supported by this conversion, but in most cases we should fail fast when a client attempts to use the unsupported functionality.

### CODE REVIEW

File: src/server/pfs/server/source.go [hunk 1]
```diff
@@ -1,93 +1,18 @@
...
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset"
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
-	"github.com/pachyderm/pachyderm/src/server/pkg/tar"
 	"golang.org/x/net/context"
 )
 
-// FileReader is a PFS wrapper for a fileset.MergeReader.
-// The primary purpose of this abstraction is to convert from index.Index to
-// pfs.FileInfo and to convert a set of index hashes to a file hash.
-type FileReader struct {
```
Comment (brendoncarroll) [inline src/server/pfs/server/source.go:20]: 👍

File: src/server/pfs/server/source.go [hunk 2]
```diff
@@ -119,38 +44,50 @@ func (s *source) Iterate(ctx context.Context, cb func(*pfs.FileInfo, fileset.Fil
...
+				fi.FileType = computedFi.FileType
+				fi.SizeBytes = computedFi.SizeBytes
+				fi.Hash = computedFi.Hash
 			}
-			fi.Hash = hashBytes
-		}
-		if err := cb(fi, fr); err != nil {
-			return err
 		}
-		delete(cache, idx.Path)
```
Comment (brendoncarroll) [inline src/server/pfs/server/source.go:144]: Doesn't lexicographical order ensure that a parent is emitted before it's child?  By the time anything is emitted it's FileInfo is no longer needed by any path coming later.
Why can't we delete here?

File: src/server/pfs/server/val_server.go [hunk 3]
```diff
@@ -19,16 +19,61 @@ type validatedAPIServer struct {
...
-func newValidatedAPIServer(inner APIServer, env *serviceenv.ServiceEnv) *validatedAPIServer {
+func newValidatedAPIServer(embeddedServer APIServer, env *serviceenv.ServiceEnv) *validatedAPIServer {
 	return &validatedAPIServer{
-		APIServer: inner,
+		APIServer: embeddedServer,
 		env:       env,
 	}
 }
 
+// DeleteRepoInTransaction is identical to DeleteRepo except that it can run
```
Comment (brendoncarroll) [inline src/server/pfs/server/val_server.go:29]: 👍

File: src/server/pfs/server/source.go (hunk 2 above)
Comment (brycemcanally) [reply src/server/pfs/server/source.go:144]: Yes, but for the subsequent child files we need to check if the parent directory info is cached. This allows us to correctly avoid progressing the second stream when it has already been progressed past a file. When we compute a directory info we progress the second stream past all of the files / directories in that directory.

Comment (brycemcanally) [reply src/server/pfs/server/source.go:144]: We can delete a directory when we are past all of its files / directories, I just need to put more thought into how we handle that. I'm not too concerned about this caching for now, which is why I left it as a TODO.

**Pull request merged**